mesonbuild/interpreter.py

```diff
@@ -2386,24 +2386,21 @@ def add_languages(self, args, required):
             mlog.log('Native %s compiler: ' % comp.get_display_language(), mlog.bold(' '.join(comp.get_exelist())), version_string, sep='')
 
             # If <language>_args/_link_args settings are given on the
-            # command line, use them.
+            # command line, use them, otherwise take them from env.
+            (preproc_args, compile_args, link_args) = environment.get_args_from_envvars(comp)
             for optspec in self.build.environment.cmd_line_options.projectoptions:
                 (optname, optvalue) = optspec.split('=', maxsplit=1)
-                if optname.endswith('_link_args'):
-                    lang = optname[:-10]
-                    self.coredata.external_link_args.setdefault(lang, []).append(optvalue)
+                if optname == lang + '_link_args':
+                    link_args = shlex.split(optvalue)
                 elif optname.endswith('_args'):
-                    lang = optname[:-5]
-                    self.coredata.external_args.setdefault(lang, []).append(optvalue)
-            # Otherwise, look for definitions from environment
-            # variables such as CFLAGS.
-            (preproc_args, compile_args, link_args) = environment.get_args_from_envvars(comp)
-            if not comp.get_language() in self.coredata.external_preprocess_args:
-                self.coredata.external_preprocess_args[comp.get_language()] = preproc_args
-            if not comp.get_language() in self.coredata.external_args:
-                self.coredata.external_args[comp.get_language()] = compile_args
-            if not comp.get_language() in self.coredata.external_link_args:
-                self.coredata.external_link_args[comp.get_language()] = link_args
+                    compile_args = shlex.split(optvalue)
```

Comment:
I think `-Dlang_args` should append to (override) the values from the environment, not replace them.

Reply:
In 0.46.0 it replace the env, but then append when you pass -Dlang_args multiple times.

I personally don't think it should append to env, they are 2 incompatible way of setting compiler args, like in which order? Same question when using (sub)project(default_options=['c_args=-foo']). It is clear for all options that they override values, I don't think c_args is any different.

Reply:
Am I misreading the diff or was it the other way around? The default values were set with `-Dlang_args` and they were overriden by the environment variable values.

Reply:
It sets value from env only if lang is not in the dict, and if we have a CMD line it gets added in the dict